cs3api4lab/locks/cs3.py

```diff
@@ -0,0 +1,95 @@
+import time
+
+from cs3api4lab.locks.base import LockBase
+from cs3api4lab.utils.file_utils import FileUtils
+from cs3api4lab.api.cs3_user_api import Cs3UserApi
+from cs3api4lab.exception.exceptions import FileLockedError
+
+import cs3.storage.provider.v1beta1.provider_api_pb2 as storage_api
+import cs3.types.v1beta1.types_pb2 as cs3_types
+import cs3.identity.user.v1beta1.resources_pb2 as id_res
+import cs3.storage.provider.v1beta1.resources_pb2 as storage_resources
+import cs3.rpc.v1beta1.code_pb2 as cs3code
+
+import google.protobuf.json_format as json_format
+
+
+class Cs3(LockBase):
+
+    def __init__(self, log, config):
+        super().__init__(log, config)
+        self.user_api = Cs3UserApi(log)
+        self.config = config
+
+    def handle_locks(self, file_path, endpoint):
+        ref = FileUtils.get_reference(file_path, endpoint)
+        lock = self._get_lock(ref)
+        if not lock:
+            self._set_lock(ref)
+        elif self._is_lock_mine(lock):
+            self._refresh_lock(ref)
+        elif not self._is_lock_expired(lock):
+            self._unlock(ref, lock)
+            self._set_lock(ref)
+        else:
+            raise FileLockedError("File %s is locked" % file_path)
+
+    def is_valid_external_lock(self, file_path, endpoint):
+        ref = FileUtils.get_reference(file_path, endpoint)
+        lock = self._get_lock(ref)
+        if lock:
+            return self._is_lock_expired(lock) and not self._is_lock_mine(lock)
+        return False
+
+    @staticmethod
+    def _is_lock_expired(lock):
```

Comment:
This method seems to be both not needed (the storage layer never returns an expired lock) and incorrect, as the lock internal properties are not accessible like this. You could compare with the wopiserver implementation at https://github.com/cs3org/wopiserver/blob/master/src/core/cs3iface.py#L215-L226

Reply:
The lock is mapped to json first, so the props are accessible this way. I will take a look if it's needed anyway